### PR TITLE
Make brass catcher a container and catch brass

### LINKED ISSUE #42313: Brass Catcher Does Not Actually Collect Spent Rounds
# Describe the bug
No crash or error report or any message here really, but I've noticed that the brass catcher does not actually collect the casings that are fired from a firearm. Specifically, I have been using a reclaimed M16A4 with 5.56 NATO M855A1. I have fired probably over 1000 rounds with this gun and have had the brass catcher on for most of those, yet I have seen none of those rounds appear in my inventory or the brass catcher. I even disassembled it thinking I was just being dumb, but the only items yielded from the disassembly were rags and thread. There's no bullet casings to be found. I'm not sure where they are going.

# Steps To Reproduce
1) Get firearm.
2) Put brass catcher on firearm.
3) Shoot the bad guys.
4) The casings are nowhere to be found??

# Expected behavior
Spent casings to be located in the player inventory or brass catcher or to appear on the ground or player inventory upon disassembly of the brass catcher.

# Screenshots
n/a

# Versions and configuration
- OS: Windows
    - OS Version: 10.0 1903
- Game Version: 0.E-4297-g3c67e80 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Alternative Map Key [alt_map_key],
    My Sweet Cataclysm [my_sweet_cataclysm],
    No Fungal Monsters [No_Fungi],
    SpeedyDex [speedydex],
    Stats Through Skills [StatsThroughSkills]
]

# Additional context
n/a

### PR DESCRIPTION
**Pull request by mqrause**
#### Summary

SUMMARY: Bugfixes "Brass catcher catches casings"

#### Purpose of change

Fixes #42313

#### Describe the solution

Added BRASS_CATCHER flag to identify mods capable of catching brass.
Brass catchers need a container pocket so a weapon can insert casings into it. If there's no pocket or it's full/too small, casings will be dropped to the ground.
When a brass catcher with contents is installed in a weapon, `U`nloading will allow you to select the brass catcher for unloading.

#### Describe alternatives you've considered

I tried a few things. Simply changing `CONTAINER` to `MOD` in the original ranged.cpp would make the catcher work. Casing would then be released upon removing it. This would however mean there's an infinite amount of casings you can store in the weapon. And I never tested if saving and loading would work for that.

Another thing was making the brass catcher add a container pocket to the gun. This had some weird behavior like not being able to select the casings in the inventory despite them being displayed there. You also had to unload the weapon to remove the casings and removing the brass catcher would delete the casings in the weapon (I assume this is related to the bug I tried to fix in #42088, when removing the casings from the wielded weapon, it inserts them back into it, but then the entire pocket gets removed).

To overcomplicate things I then tried to both make the brass catcher a container and have it add a container to the weapon. Upon installing and removing the contents would then be moved from the mod to the weapon and vice versa. This would theoretically give the ability to access the casings without removing the brass catcher, but I didn't manage to move the items on installing the brass catcher and then gave up on it.

#### Testing

Shoot weapon with and without brass catcher.

#### Additional context

Someone with actual gun knowledge should probably look at the brass catcher stats. I based the pocket data on existing item volume, some pictures and other containers with similar sizes.

### CODE REVIEW

File: data/json/items/gunmod/brass_catcher.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
     "color": "light_gray",
     "location": "brass catcher",
     "mod_targets": [ "shotgun", "smg", "rifle", "pistol", "launcher" ],
+    "pocket_data": [ { "pocket_type": "CONTAINER", "rigid": true, "max_contains_volume": "249 ml", "max_contains_weight": "1 kg" } ],
```
Comment (Qrox) [inline data/json/items/gunmod/brass_catcher.json:16]: According to the item descript and crafting recipe, brass catcher is a bag made from rags, so it's probably not rigid.

File: src/ranged.cpp [hunk 2]
```diff
@@ -1558,9 +1558,9 @@ static void cycle_action( item &weap, const tripoint &pos )
 
     if( weap.ammo_data() && weap.ammo_data()->ammo->casing ) {
         const itype_id casing = *weap.ammo_data()->ammo->casing;
-        if( weap.has_flag( "RELOAD_EJECT" ) || weap.gunmod_find( itype_brass_catcher ) ) {
-            weap.put_in( item( casing ).set_flag( "CASING" ), item_pocket::pocket_type::CONTAINER );
-        } else {
+        item *brass_catcher = weap.gunmod_find( itype_brass_catcher );
+        if( !( weap.has_flag( "RELOAD_EJECT" ) || ( brass_catcher && brass_catcher->put_in( item( casing ),
+                item_pocket::pocket_type::CONTAINER ).success() ) ) ) {
```
Comment (Qrox) [inline src/ranged.cpp:1563]: I'm not sure if this logic is correct. If a weapon has the "RELOAD_EJECT" flag, before this change it would put the casing in the weapon, but after this it will do nothing, making the casing disappear.

Also, later at line 1579-1581 ammo linkages are also added to brass catchers, which should also be updated.

File: data/json/items/gunmod/brass_catcher.json (hunk 1 above)
Comment (mqrause) [reply data/json/items/gunmod/brass_catcher.json:16]: Yeah, I really wasn't sure about that one. 250ml seems a bit much for just a bag, since then it would have a few liters of volume.

File: src/ranged.cpp (hunk 2 above)
Comment (mqrause) [reply src/ranged.cpp:1563]: Oops, I guess I confused myself with prior changes to that. Not that putting casings into the weapon would have worked before, though. Weapons don't have any container pocket, so it just did nothing. I'll separate the "RELOAD_EJECT" logic and leave it as it was before for another fix.

File: src/ranged.cpp [hunk 3]
```diff
@@ -1576,9 +1578,8 @@ static void cycle_action( item &weap, const tripoint &pos )
     const auto mag = weap.magazine_current();
     if( mag && mag->type->magazine->linkage ) {
         item linkage( *mag->type->magazine->linkage, calendar::turn, 1 );
-        if( weap.gunmod_find( itype_brass_catcher ) ) {
-            linkage.set_flag( "CASING" );
-            weap.put_in( linkage, item_pocket::pocket_type::CONTAINER );
+        if( brass_catcher ) {
+            brass_catcher->put_in( linkage, item_pocket::pocket_type::CONTAINER );
```
Comment (Qrox) [inline src/ranged.cpp:1582]: If the brass catcher is full the linkage will not be added to the ground.

File: src/ranged.cpp [hunk 4]
```diff
@@ -1576,9 +1578,8 @@ static void cycle_action( item &weap, const tripoint &pos )
     const auto mag = weap.magazine_current();
     if( mag && mag->type->magazine->linkage ) {
         item linkage( *mag->type->magazine->linkage, calendar::turn, 1 );
-        if( weap.gunmod_find( itype_brass_catcher ) ) {
-            linkage.set_flag( "CASING" );
```
Comment (Qrox) [inline src/ranged.cpp:1580]: Why is this line removed?

File: src/ranged.cpp [hunk 5]
```diff
@@ -1556,11 +1556,13 @@ static void cycle_action( item &weap, const tripoint &pos )
...
     }
 
+    item *brass_catcher = weap.gunmod_find( itype_brass_catcher );
     if( weap.ammo_data() && weap.ammo_data()->ammo->casing ) {
         const itype_id casing = *weap.ammo_data()->ammo->casing;
-        if( weap.has_flag( "RELOAD_EJECT" ) || weap.gunmod_find( itype_brass_catcher ) ) {
+        if( weap.has_flag( "RELOAD_EJECT" ) ) {
             weap.put_in( item( casing ).set_flag( "CASING" ), item_pocket::pocket_type::CONTAINER );
-        } else {
+        } else if( !( brass_catcher && brass_catcher->put_in( item( casing ),
```
Comment (Qrox) [inline src/ranged.cpp:1564]: Previously it sets the "CASING" flag, but now it does not.

File: src/ranged.cpp [hunk 6]
```diff
@@ -1556,11 +1556,13 @@ static void cycle_action( item &weap, const tripoint &pos )
         cargo = vp->vehicle().get_parts_at( pos, "CARGO", part_status_flag::any );
     }
 
+    item *brass_catcher = weap.gunmod_find( itype_brass_catcher );
```
Comment (KorGgenT) [inline src/ranged.cpp:1559]: use a flag instead of a specific itype.

File: src/ranged.cpp (hunk 4 above)
Comment (mqrause) [reply src/ranged.cpp:1580]: Previously, this was used to find them and add the to inventory/ground when removing the brass catcher since they were not containers. Now the items are directly put into the brass catcher and removing it gives a filled brass catcher that doesn't need additional evaluation, so from my understanding it's unnecessary.

File: src/ranged.cpp (hunk 6 above)
Comment (mqrause) [reply src/ranged.cpp:1559]: So give a flag to brass catcher and make a gunmod_find alternative that looks for flags?

File: src/ranged.cpp (hunk 4 above)
Comment (Qrox) [reply src/ranged.cpp:1580]: I see, thanks.

**Comment by kevingranade** [comment]
IIRC what we used to do is have unload of the gun also trigger unload of the brass catcher.
Ideally we could select the mod itself and unload it directly, but that might be a lot of infrastructure for negligible benefit.

**Comment by mqrause** [comment]
> Ideally we could select the mod itself and unload it directly, but that might be a lot of infrastructure for negligible benefit.

Turns out that's already possible. It's a little awkward, though, since you can still unload an already empty weapon with a filled brass catcher. And it's missing an indicator in the UI if the brass catcher contains anything.

**Comment by ZhilkinSerg** [comment]
Can you rebase on top of current master please?

**Comment by mqrause** [comment]
This probably didn't go as intended.

**Comment by ZhilkinSerg** [comment]
Needs some handling of the case when there is no empty space in brass catcher, otherwise you will get an error each time you shoot (empty shell will correctly drop on the ground though).

```
 DEBUG    : tried to put an item (shot_hull) count (1) in a container (brass_catcher) that cannot contain it: pocket unacceptable because not enough space

 FUNCTION : ret_val<bool> item::put_in(const item&, item_pocket::pocket_type)
 FILE     : src/item.cpp
 LINE     : 1102
```

**Comment by ZhilkinSerg** [comment]
Not necessary cause by this PR, but following error appears when you try to put a stack of more than 20 shotgun hulls into brass catcher:

```
 DEBUG    : tried to put an item (shot_hull) count (38) in a container (brass_catcher) that cannot contain it: pocket unacceptable because item too big

 FUNCTION : ret_val<bool> item::put_in(const item&, item_pocket::pocket_type)
 FILE     : src/item.cpp
 LINE     : 1102
```
